### DISCUSSION TITLE
Performance & Slowness

### DISCUSSION THREAD

**Discussion by thewh1teagle** [tazz4843/whisper-rs#111] [General]
I tried to run `whisper-rs-cli` on `16khz` `mono` wav file with length of `30s` (default cli params)
It took `1m` and `9.408s`.

Spec: `amd ryzen 5 4500u`, `windows 11`

1. It looks pretty slow to me, is there a way to improve it? 
2. What's the best way to stream data into the input instead of holding everything in ram at a time?
I tried to split the samples and call `state.full_n_segments` multiple times and it was, slower, took `1m` and `47.975s`

**Comment by tazz4843** [comment]
Best question is are you compiling in release mode?

Failing that, it'd be nice to see your code.

There's no way to stream audio due to fundamental limitations in the whisper model itself, which feeds audio in 30-second chunks, so you must load the entire file into memory and process that.

**Reply by thewh1teagle** [reply]
Very good question, I didn't tried yet in release mode, missed it
https://github.com/thewh1teagle/ruscribe

if whisper feeds audio in 30 second chunks, 
we can stream 30 seconds chunk each time? no? (at least that instead of load the entire file)

**Reply by tazz4843** [reply]
For feeding in 30 second chunks, I found the following issues on upstream's repo:

https://github.com/ggerganov/whisper.cpp/issues/137
https://github.com/ggerganov/whisper.cpp/issues/426
https://github.com/ggerganov/whisper.cpp/issues/1428

This would need upstream support unfortunately before we have any way to implement it. Right now we expose the entirety of whisper.cpp's API.

**Reply by thewh1teagle** [reply]
I'll see them.
Wondering if it will slow down the process.
I think it's not usable to use it when you load everything to memory first. I have to load it in chunks of 10-30 seconds from fs,
But I hope that it won't affect performance as everytime I `fill` the model with new sample it will do some heavy stuff,
any toughts?

**Comment by thewh1teagle** [comment]
Checked in release mode
The wav file with length of 30 seconds took
`0m` and `42.074s`, better!

**Reply by tazz4843** [reply]
It's still rather interesting that it takes so long. Perhaps you could use something like `cargo flamegraph` to see where the issue lies.

https://github.com/flamegraph-rs/flamegraph

**Comment by thewh1teagle** [comment]
1. Can we hide somehow the output of whisper.cpp completly? (it always show `loading model ....` )?
2. I tried to use progress callback for using with [indicatif](https://github.com/console-rs/indicatif)
When I set it inside the closure of `on_progress` with `move || {}` I get `access violation error` (`core dump`)

**Reply by tazz4843** [reply]
> Can we hide somehow the output of whisper.cpp completly? (it always show loading model .... )?

I've got a logging trampoline function set up locally, if you like I can push it as a utility and set it up for you.

> I tried to use progress callback for using with [indicatif](https://github.com/console-rs/indicatif)
> When I set it inside the closure of on_progress with move || {} I get access violation error (core dump)

This sounds like a bug, can you open an issue with some code samples?

**Reply by tazz4843** [reply]
See #112 for some logging utilities that will help avoid those output lines

**Reply by thewh1teagle** [reply]
@tazz4843 
![image](https://github.com/tazz4843/whisper-rs/assets/61390950/e9bf0b6b-2684-4c9c-8cef-70806f6431cd)

**Reply by thewh1teagle** [reply]
Also with tracing feature
![image](https://github.com/tazz4843/whisper-rs/assets/61390950/08ac3358-4997-4144-bdcb-a5b31f7baf8b)

**Reply by tazz4843** [reply]
Should be fixed, give it another shot.

**Reply by thewh1teagle** [reply]
Going to try.
Do you have idea about this? https://github.com/tazz4843/whisper-rs/issues/114

**Comment by thewh1teagle** [comment]
I made this open source project [vibe](https://github.com/thewh1teagle/vibe)
I'm planning to show the text transcribed in realtime and progress bar.
<video src="https://github.com/tazz4843/whisper-rs/assets/61390950/6decf4c3-5bb9-46d2-8aa9-5e512171ab86">
